pallets/frequency-tx-payment/src/lib.rs

```diff
@@ -0,0 +1,428 @@
+	pub fn length_to_fee(length: u32) -> BalanceOf<T> {
+		T::LengthToFee::weight_to_fee(&Weight::from_ref_time(length as u64))
+	}
+
+	/// Compute the unadjusted portion of the weight fee by invoking the configured `WeightToFee`
+	/// impl. Note that the input `weight` is capped by the maximum block weight before computation.
+	pub fn weight_to_fee(weight: Weight) -> BalanceOf<T> {
+		// cap the weight to the maximum defined in runtime, otherwise it will be the
+		// `Bounded` maximum of its data type, which is not desired.
+		let capped_weight = weight.min(T::BlockWeights::get().max_block);
+		T::WeightToFee::weight_to_fee(&capped_weight)
+	}
+}
+
+/// Custom Transaction Validity Errors for ChargeFrqTransactionPayment
+pub enum ChargeFrqTransactionPaymentError {
+	/// The call is not eligible to be paid for with Capacity
+	CallIsNotCapacityEligible,
+	/// The account key is not associated with an MSA
+	InvalidMsaKey,
+	/// The Capacity Target does not exist
+	TargetCapacityNotFound,
+}
+
+/// Require the transactor pay for themselves and maybe include a tip to gain additional priority
+/// in the queue.
+///
+/// # Transaction Validity
+///
+/// This extension sets the `priority` field of `TransactionValidity` depending on the amount
+/// of tip being paid per weight unit.
+///
+/// Operational transactions will receive an additional priority bump, so that they are normally
+/// considered before regular transactions.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct ChargeFrqTransactionPayment<T: Config>(#[codec(compact)] BalanceOf<T>);
+
+impl ChargeFrqTransactionPaymentError {
+	pub fn into(self) -> TransactionValidityError {
+		TransactionValidityError::from(InvalidTransaction::Custom(self as u8))
+	}
+}
+
+impl<T: Config> ChargeFrqTransactionPayment<T>
+where
+	BalanceOf<T>: Send + Sync + FixedPointOperand + IsType<CapacityBalanceOf<T>> + From<u64>,
+	<T as frame_system::Config>::RuntimeCall:
+		Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo> + IsSubType<Call<T>>,
+{
+	/// Utility construct from tip.
+	pub fn from(tip: BalanceOf<T>) -> Self {
+		Self(tip)
+	}
+
+	/// Return the tip as being chosen by the transaction sender.
+	pub fn tip(&self, call: &<T as frame_system::Config>::RuntimeCall) -> BalanceOf<T> {
+		match call.is_sub_type() {
+			Some(Call::pay_with_capacity { .. }) => Zero::zero(),
+			_ => self.0,
+		}
+	}
+
+	/// Withdraws fee from either Capacity ledger or Token account.
+	fn withdraw_fee(
+		&self,
+		who: &T::AccountId,
+		call: &<T as frame_system::Config>::RuntimeCall,
+		info: &DispatchInfoOf<<T as frame_system::Config>::RuntimeCall>,
+		len: usize,
+	) -> Result<(BalanceOf<T>, InitialPayment<T>), TransactionValidityError> {
+		match call.is_sub_type() {
+			Some(Call::pay_with_capacity { call }) => {
+				let extrinsic_weight = T::CapacityCalls::get_stable_weight(call)
```

Comment:
The non-stable weights are still used to track how much weight is consumed in the block accurately. So if we include any stable weight stuff here, it would interfere with keeping track of the actual weight consumption for a block.

But I might need clarification on what you are suggesting. Let's talk about it more.

Reply:
I didn't mean the stable weights. My suggestion was to do a benchmark for this signed extension and use that generated weight in the pay_with_capacity weights to have a more accurate weight for those extrinsic but let's talk about it if not clear.

Reply:
Ah, I think I know what you mean. Are you saying to push stuff related to Capacity here instead of the overhead? I have not had a chance to look into too much, but yes, you are right. That is an option. I've had thoughts about creating a different extrinsic to do some checks with MSA accounts since we do those checks on every extrinsic. However, I have not had a chance to dive into the different options.